### ISSUE TITLE
Can't Chromecast

### ISSUE THREAD

**Issue by CyanJDroid** [MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-#1265]
**OS:**
Antergos (Arch) Linux

**Issue Description:**
Unless I'm just blind, I don't see any option to Chromecast, despite the latest version (3.3.0, which I have double checked to ensure I have) stating it now supports it. I've checked to ensure my Chromecasts are visible on my network, which they are, and I can cast to them just fine from my phone and from Chrome on the same computer. Not sure if this is a bug, I'm doing something wrong, or if this is only an issue I'm having.

**Comment by jostrander** [comment]
What operating system and do you have bonjour or avahi installed?

**Comment by jostrander** [comment]
What operating system and do you have bonjour or avahi installed?
On Jul 3, 2016 7:32 PM, "CyanJDroid" <notifications@github.com> wrote:

*OS:*
Antergos (Arch) Linux

*Issue Description:*
Unless I'm just blind, I don't see any option to Chromecast, despite the
latest version (3.3.0, which I have double checked to ensure I have)
stating it now supports it. I've checked to ensure my Chromecasts are
visible on my network, which they are, and I can cast to them just fine
from my phone and from Chrome on the same computer. Not sure if this is a
bug, I'm doing something wrong, or if this is only an issue I'm having.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-/issues/1265>,
or mute the thread
<https://github.com/notifications/unsubscribe/AAucnY_C_uCOH0sEVHpojGE-iEA464JOks5qSEalgaJpZM4JD-MF>
.

**Comment by CyanJDroid** [comment]
Antergos (Arch) Linux, and yes Avahi is installed on my system.

**Comment by jostrander** [comment]
In theory it should be working, are you using a chrome cast or a cast
compatible device?

**Comment by CyanJDroid** [comment]
I have two Chromecasts and one Android TV device on my network, so a mixture of both.

**Comment by jostrander** [comment]
How did you install the update?

**Comment by CyanJDroid** [comment]
Using my system package manager.

I did a fresh install of the app on another computer that's never had it installed on it before (also running Antergos) to see if it'd work there, and it doesn't, so a fresh install doesn't fix it. I also tried it on a Windows machine on my network and while it does work there, it fails to detect the Android TV (which isn't that big of a problem for me, but still seems like something that should work).

**Comment by jostrander** [comment]
That could be just a discovery issue, the old one probably uses cast v1
dial discovery. Which we don't have support but in for. As far as it not
working via package manager... can you try installing from source / dev
environment? It could possibly provide more insight as to what's going on.

**Comment by jostrander** [comment]
Also, you don't have a firewall or anything preventing network discovery
from UDP port 5353? I believe that's where it comes from.

**Comment by CyanJDroid** [comment]
I ensured that my firewall isn't blocking any activity on port 5353, and still nothing. Even tried turning the firewall off all together, but to no avail. Considering I can still cast from Chrome I doubt it's a firewall issue. Followed the directions for installing from source in the tar.gz's readme file, but still no casting. Looking through the terminal output, it's complaining a lot that the app is using "Apple Bonjour compatibility layer of Avahi" and wants the app to "use the native API of Avahi" instead. It also complains that "DNSServiceRegister() is not supported (or only supported partially) in the Apple Bonjour compatibility layer of Avahi".

**Comment by MarshallOfSound** [comment]
@CyanJDroid Those `avahi` warnings are completely normal.  If I had to guess, there is something missing architecturally from your networking stack that is preventing avahi from scanning correctly.  (Just a guess).

Without your specific hardware and driver configuration it is virtually impossible for @jostrander or me to figure out exactly what is going wrong.  I have personally tested Chromecast support on W7, W8, W10, macOS, Ubuntu 14 and Ubuntu 16.

Purely due to hard drive space we can't have VM's for all the OS's under the sun 😆 

The final thing I can get you to try is to run the following bash script.

```bash
cd ~/
mkdir tmp
cd tmp
touch test.js
npm i mdns
vim test.js
```

And input the following into `test.js`

```js
const mdns = require('mdns');

var browser = mdns.createBrowser(mdns.tcp('googlecast'));
browser.on('serviceUp', function(service) {
  console.log("service up: ", service);
});
browser.on('serviceDown', function(service) {
  console.log("service down: ", service);
});
browser.start();
```

Then run  

```bash
node test.js
```

And see if any services pop up 👍

**Comment by jostrander** [comment]
You literally beat me to the bunch by 3 seconds.

**Comment by CyanJDroid** [comment]
@MarshallOfSound This is the output I get from that:
`*** WARNING *** The program 'node' uses the Apple Bonjour compatibility layer of Avahi.
*** WARNING *** Please fix your application to use the native API of Avahi!
*** WARNING *** For more information see <http://0pointer.de/avahi-compat?s=libdns_sd&e=node>
*** WARNING *** The program 'node' called 'DNSServiceRegister()' which is not supported (or only supported partially) in the Apple Bonjour compatibility layer of Avahi.
*** WARNING *** Please fix your application to use the native API of Avahi!
*** WARNING *** For more information see <http://0pointer.de/avahi-compat?s=libdns_sd&e=node&f=DNSServiceRegister>
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: getaddrinfo -3008
    at errnoException (/home/cyanjdroid/tmp/node_modules/mdns/lib/resolver_sequence_tasks.js:199:11)
    at getaddrinfo_complete (/home/cyanjdroid/tmp/node_modules/mdns/lib/resolver_sequence_tasks.js:112:10)
    at GetAddrInfoReqWrap.oncomplete (/home/cyanjdroid/tmp/node_modules/mdns/lib/resolver_sequence_tasks.js:120:9)
`

@jostrander This is the output I get from that:
`/home/cyanjdroid/node-mdns-easy/dist/index.js:5
var mdns = new MDns();
           ^

TypeError: MDns is not a constructor
    at Object.<anonymous> (/home/cyanjdroid/node-mdns-easy/dist/index.js:5:12)
    at Module._compile (module.js:541:32)
    at Object.Module._extensions..js (module.js:550:10)
    at Module.load (module.js:458:32)
    at tryModuleLoad (module.js:417:12)
    at Function.Module._load (module.js:409:3)
    at Module.runMain (module.js:575:10)
    at run (node.js:348:7)
    at startup (node.js:140:9)
    at node.js:463:3
`

**Comment by MarshallOfSound** [comment]
@CyanJDroid Change @jostrander's code to be `new MDns.default()`

**Comment by CyanJDroid** [comment]
Now I get this:

```
*** WARNING *** The program 'node' uses the Apple Bonjour compatibility layer of Avahi.
*** WARNING *** Please fix your application to use the native API of Avahi!
*** WARNING *** For more information see <http://0pointer.de/avahi-compat?s=libdns_sd&e=node>
*** WARNING *** The program 'node' called 'DNSServiceRegister()' which is not supported (or only supported partially) in the Apple Bonjour compatibility layer of Avahi.
*** WARNING *** Please fix your application to use the native API of Avahi!
*** WARNING *** For more information see <http://0pointer.de/avahi-compat?s=libdns_sd&e=node&f=DNSServiceRegister>
{ Error: getaddrinfo -3008
    at errnoException (/home/cyanjdroid/node-mdns-easy/node_modules/mdns/lib/resolver_sequence_tasks.js:199:11)
    at getaddrinfo_complete (/home/cyanjdroid/node-mdns-easy/node_modules/mdns/lib/resolver_sequence_tasks.js:112:10)
    at GetAddrInfoReqWrap.oncomplete (/home/cyanjdroid/node-mdns-easy/node_modules/mdns/lib/resolver_sequence_tasks.js:120:9) code: -3008, errno: -3008, syscall: 'getaddrinfo' }
{ Error: getaddrinfo -3008
    at errnoException (/home/cyanjdroid/node-mdns-easy/node_modules/mdns/lib/resolver_sequence_tasks.js:199:11)
    at getaddrinfo_complete (/home/cyanjdroid/node-mdns-easy/node_modules/mdns/lib/resolver_sequence_tasks.js:112:10)
    at GetAddrInfoReqWrap.oncomplete (/home/cyanjdroid/node-mdns-easy/node_modules/mdns/lib/resolver_sequence_tasks.js:120:9) code: -3008, errno: -3008, syscall: 'getaddrinfo' }
{ Error: getaddrinfo -3008
    at errnoException (/home/cyanjdroid/node-mdns-easy/node_modules/mdns/lib/resolver_sequence_tasks.js:199:11)
    at getaddrinfo_complete (/home/cyanjdroid/node-mdns-easy/node_modules/mdns/lib/resolver_sequence_tasks.js:112:10)
    at GetAddrInfoReqWrap.oncomplete (/home/cyanjdroid/node-mdns-easy/node_modules/mdns/lib/resolver_sequence_tasks.js:120:9) code: -3008, errno: -3008, syscall: 'getaddrinfo' }
```

**Comment by MarshallOfSound** [comment]
@jostrander I thought our resolver sequence fixed this issue?  It looks familiar?

**Comment by jostrander** [comment]
I may have changed the resolver to only try and use IPv4 addressing... 

https://github.com/GPMDP/node-mdns-easy/blob/master/src/browser.js#L17

change `families[4]` to `families[0]` and rerun `npm run compile` and then rerun `node dist/index.js`. You may have to change the code again to MDns().default() like @MarshallOfSound suggested. 

Are you using IPv6 addressing by chance?

**Comment by CyanJDroid** [comment]
Same results, though I will note that each of those errors appears to correspond to a cast device, because only two of my cast devices are online this time and only two of the errors showed up, whereas last time all three of my cast devices were online and three showed up.

I am not using IPv6 addressing.

**Comment by CyanJDroid** [comment]
Ah! I suppose I should have mentioned that all of my Chromecast devices are Chromecast v1! Based on that plus that other issue, it would appear that is our problem!

**Comment by Mossop** [comment]
I am running on Windows with two cast v2 devices and neither of them are picked up.

**Comment by Mossop** [comment]
@Jostrander No I do not

**Comment by jostrander** [comment]
Can you install bonjour and restart GPMDP and see if it resolves your issue?

**Comment by Mossop** [comment]
Ok that fixed it

**Comment by jostrander** [comment]
Sorry for the inconvenience, our fallback library (that doesn't require bonjour) is super flaky.

**Comment by Alipoodle** [comment]
I'm also experiencing some problems! I can't see the button in the first place.
Although i'm using Sonos speakers. They do show up on a computers and mobile for casting!

**Comment by MarshallOfSound** [comment]
> Can you install bonjour and restart GPMDP and see if it resolves your issue?

**Comment by Alipoodle** [comment]
From what i remember i have Bonjour, although i will install it, and see back!

**Comment by Alipoodle** [comment]
Still isn't showing! Just to check this is Apple's Bonjour!

**Comment by MarshallOfSound** [comment]
Can you follow the instructions here

https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-/issues/1292#issuecomment-231511635

**Comment by Alipoodle** [comment]
Added to that page ^

**Comment by MarshallOfSound** [comment]
This issue thread appears to have multiple resolved issues and one unresolved issue probably related to a OS specific network stack so I'm going to close this out

**Comment by njam** [comment]
When I'm running the following test script with mdns on Arch Linux:
```js
const mdns = require('mdns');

var sequence = [
  mdns.rst.DNSServiceResolve(),
  mdns.rst.getaddrinfo({ families: [0] }),
  mdns.rst.makeAddressesUnique(),
];
var browser = mdns.createBrowser(mdns.tcp('googlecast'), { resolverSequence: sequence });
browser.on('serviceUp', (service) => { console.log(service) });
browser.start();
``` 

I also get error "-3008":
```
Error: getaddrinfo -3008
    at errnoException (/tmp/tmp.TYjbuG604r/node_modules/mdns/lib/resolver_sequence_tasks.js:200:11)
    at getaddrinfo_complete (/tmp/tmp.TYjbuG604r/node_modules/mdns/lib/resolver_sequence_tasks.js:112:10)
    at GetAddrInfoReqWrap.oncomplete (/tmp/tmp.TYjbuG604r/node_modules/mdns/lib/resolver_sequence_tasks.js:121:9)
``` 

The address that is being resolved is `ea317ef7-2c7c-d4ec-d8cf-5c5ca034026c.local.`.
Any hints about how to continue debugging? I can't even find any reference to error "-3008" from "getaddrinfo".

---

I'm not 100% sure this is the reason that the Chromecast doesn't show up for me in GPMDP. When I'm starting GPMDP from the command line I don't see any error like the "-3008". Is there a better way to debug this in GPMDP?
